src/artm/regularizer_sandbox/label_regularization_phi.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright 2014, Additive Regularization of Topic Models.
+
+// Author: Murat Apishev (great-mel@yandex.ru)
+
+#include "artm/regularizer_sandbox/label_regularization_phi.h"
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "artm/core/protobuf_helpers.h"
+#include "artm/core/regularizable.h"
+#include "artm/core/topic_model.h"
+
+namespace artm {
+namespace regularizer_sandbox {
+
+bool LabelRegularizationPhi::RegularizePhi(::artm::core::Regularizable* topic_model, double tau) {
+  // read the parameters from config and control their correctness
+  const int topic_size = topic_model->topic_size();
+  const int token_size = topic_model->token_size();
+
+  std::vector<bool> topics_to_regularize;
+  if (config_.topic_name().size() == 0)
+    topics_to_regularize.assign(topic_size, true);
+  else
+    topics_to_regularize = core::is_member(config_.topic_name(), topic_model->topic_name());
+
+  bool use_all_classes = false;
+  if (config_.class_id_size() == 0) {
+    use_all_classes = true;
+  }
+
+  bool has_dictionary = true;
+  if (!config_.has_dictionary_name()) {
+    has_dictionary = false;
+  }
+
+  auto dictionary_ptr = dictionary(config_.dictionary_name());
+  if (has_dictionary && dictionary_ptr == nullptr) {
+    has_dictionary = false;
+  }
+
+  core::TokenCollectionWeights p_wt(topic_model->topic_size());
+  topic_model->FindPwt(&p_wt);
+  std::map<core::ClassId, std::vector<float> > n_t = topic_model->FindNormalizers();
+
+  // proceed the regularization
+  for (int token_id = 0; token_id < token_size; ++token_id) {
+    auto token = topic_model->token(token_id);
+    auto class_iter = n_t.find(token.class_id);
+    assert(class_iter != n_t.end());
+
+    float coefficient = 1.0f;
+    if (has_dictionary) {
+      if (use_all_classes ||
+          core::is_member(token.class_id, config_.class_id())) {
```

Comment:
if (!use_all_classes && !core_is_member(...)) continue;
This just just an alternative - your code is also OK. I just don't like extra tabulation)